### ISSUE TITLE
Audit idea: Secrets outside of environments

### ISSUE THREAD

**Issue by woodruffw** [woodruffw/zizmor#184]
### Discussed in https://github.com/woodruffw/zizmor/discussions/181

<div type='discussions-op-text'>

<sup>Originally posted by **AndrewMohawk** November 19, 2024</sup>
Loving the tool so far! I'd love to have zizmor also determine secrets referenced outside of environments within GH actions. I see it fairly often and it means that if anyone has the ability to create a PR with an updated workflow they can exfil secrets that arent tied to an environment (and that environment having branch protections). The caveat here is that an attacker would have had to submit a PR before the malicious one (usually you would see them as doing typo fixes first). </div>

**Comment by jku** [comment]
> if anyone has the ability to create a PR with an updated workflow they can exfil secrets that arent tied to an environment (and that environment having branch protections).

I agree GH secrets should be more accurately called "not-quite-public" but I also don't think it's _that_ bad: a normal pull_request event does not expose secrets to the workflow. The danger situations are:
* ability to trigger a pull_request_target event with a PR
* ability to push branches to the repo

**Comment by woodruffw** [comment]
Was thinking about this a bit more: we could probably do this without requiring any extra permissions by just checking whether each job has an environment, and flagging any secret accesses within that job. This was previously hard/annoying in zizmor because we didn't have "raw" audits/raw findings, but now we do.

**Comment by woodruffw** [comment]
(The nuance there being that the audit will only flag the absence/presence of an environment, not whether the environment has "useful" controls on it. Given that, I'm somewhat inclined to make this a pedantic-only audit to start.)
